inst/apps/connect/app.R

```diff
@@ -1120,50 +1068,391 @@ content_list_ui <- bslib::card(
-      last_updated = Sys.Date() - sample(0:365, 50, replace = TRUE)
+    tryCatch(
+      {
+        data <- chronicle_data("connect/content_list", base_path)
+
+        df <- data |> dplyr::collect()
+        if (!"date" %in% names(df) || nrow(df) == 0) {
+          return(NULL)
+        }
+
+        latest_date <- suppressWarnings(max(df$date, na.rm = TRUE))
+        df |> dplyr::filter(.data$date == latest_date)
+      },
+      error = function(e) {
+        message("Error loading content list: ", e$message)
+        NULL
+      }
+    )
+  })
+
+  # Load latest user list for owner name resolution
+  latest_user_list <- shiny::reactive({
+    tryCatch(
+      {
+        udata <- chronicle_data("connect/user_list", base_path)
+        udf <- udata |> dplyr::collect()
+        if (!"date" %in% names(udf) || nrow(udf) == 0) {
+          return(NULL)
+        }
+        latest_date <- suppressWarnings(max(udf$date, na.rm = TRUE))
+        udf |> dplyr::filter(.data$date == latest_date)
+      },
+      error = function(e) {
+        message("Error loading user list for owner resolution: ", e$message)
+        NULL
+      }
+    )
+  })
+
+  # Populate owner and type filters dynamically
+  shiny::observe({
+    data <- content_list_data()
+    if (is.null(data) || nrow(data) == 0) {
+      shiny::updateSelectInput(
+        session,
+        "content_list_environment",
+        choices = c("All"),
+        selected = "All"
+      )
+      shiny::updateSelectInput(
+        session,
+        "content_list_owner",
+        choices = c("All"),
+        selected = "All"
+      )
+      shiny::updateSelectInput(
+        session,
+        "content_list_type",
+        choices = c("All"),
+        selected = "All"
+      )
+      return()
+    }
+
+    # Detect type column
+    df <- data
+
+    # Environment choices (environment column is guaranteed)
+    env_values <- df |>
+      dplyr::pull(.data$environment) |>
+      unique()
+
+    has_env_na <- any(is.na(env_values) | env_values == "" | env_values == " ")
+    env_values <- env_values[
+      !is.na(env_values) & env_values != "" & env_values != " "
+    ] |>
+      sort()
+    if (has_env_na) {
+      env_values <- c(env_values, "(Not Set)")
+    }
+
+    shiny::updateSelectInput(
+      session,
+      "content_list_environment",
+      choices = c("All", env_values),
+      selected = "All"
+    )
+
+    # Type column on this file is app_mode; UI label remains "Type"
+    type_col <- "app_mode"
+
+    # Resolve owner names by joining latest user list on owner id
+    owners_choices <- c("All")
+    ulist <- latest_user_list()
+    if (!is.null(ulist) && nrow(ulist) > 0) {
+      # Identify join keys heuristically
+      content_owner_id_col <- intersect(
+        c(
+          "owner_id",
+          "owner_guid",
+          "content_owner_guid",
+          "user_id",
+          "owner",
+          "user_guid",
+          "guid"
```

Comment:
Is this right? Do we not _know_ what the column names are? (Or are we joining with multiple tables here and we don't know - the context is not completely obvious to me.)